### DISCUSSION TITLE
Subscriptions on AWS API Gateway Websockets with Serverless and Apollo (GraphQL)

### DISCUSSION THREAD

**Discussion by IonelLupu** [enisdenjo/graphql-ws#271] [Q&A] [answered]
I have a project where I want to implement Apollo GrapQL Subscriptions through Websockets (of course 😄 ) using the Serverless  framework on AWS. I read some issues about this here but I didn't find a solution or example about this type of setup 😢 .

I did a few things that got me closer to the setup but I still have some blockers. I did the connection part and now the client is sending this message through Websockets:
```json
{"type":"connection_init","payload":{"Authorization":"Bearer my-token"}}
```

But when working with Serveless framework on AWS we have to send (as per [docs](https://docs.aws.amazon.com/apigateway/latest/developerguide/websocket-api-develop-routes.html)) a message that contains the `action` property in order to make the entire process work and send the client message to the lambda function and handle the GraphQL query.

I am thinking I just a few things:
- a way to send a Websocket message with the `action` property because this is what "AWS API Gateway Websockets" requires. Eg: `{"action":"message","message":"here I can have my graphql subscription query"}`. Having this `action` property, AWS can trigger the lambda function that will handle the message.
- after I get the message, I need the "message interpreter" (if that's a thing) from `graphql-ws` to process it, give it to Apollo where it will return me the response. This response needs to be processed into a `graphql-ws` response so the client understand it (I assume).

But I am stuck here. Is there an example project with a `graphql-ws` server that handles the messages from the `graphql-ws` client  without express, or fastify or ws, just from a raw messaage?

**Comment by enisdenjo** [comment, marked as answer]
`graphql-ws`'s server implementation is actually just a coercing API that can be plugged into any WebSocket server library by using [`makeServer`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/modules/server.md#makeserver). Check out the built-ins for [`ws`](https://github.com/enisdenjo/graphql-ws/blob/master/src/use/ws.ts), [`uWebSockets`](https://github.com/enisdenjo/graphql-ws/blob/master/src/use/uWebSockets.ts) and [`fastify-websocket`](https://github.com/enisdenjo/graphql-ws/blob/master/src/use/fastify-websocket.ts) to get started.

Furthermore, this library is a reference implementation of the [GraphQL over WebSocket protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md) and is therefore required to have the messages format comply with the protocol; so, no, `graphql-ws` will not accept messages differing from the protocol.

However, I am aware of two libraries of intereset that aim to integrate the GraphQL over WebSocket protocol on AWS:
- [`subscriptionless`](https://github.com/andyrichardson/subscriptionless)
- [`graphql-lambda-subscriptions`](https://github.com/reconbot/graphql-lambda-subscriptions)

**Reply by IonelLupu** [reply]
I ended up using [`graphql-lambda-subscriptions`](https://github.com/reconbot/graphql-lambda-subscriptions) but I still have issues setting it up getting weird errors like `Cannot read property 'definitions' of undefined`. I will have to post an issue on their side.

Thank you very much for the reponse!

**Reply by IonelLupu** [reply]
btw. I found out why I got that error about definitions: the library was using the graphQL 16 and my app was using graphQL 15.

**Reply by SamHatoum** [reply]
@IonelLupu did you use graphql-lambda-subscriptions with or without Apollo server in the end? Thanks in advance 🙏

**Reply by IonelLupu** [reply]
@samhatoum oh. This was so long ago. I don't really remember

**Reply by SamHatoum** [reply]
I guess I'll just have to try!
